### PR TITLE
Verifier coin querry implementation

### PR DESCRIPTION
**Pull request by pvienhage**
- [ ] Tag the PR with `wip` while in development.
- [ ] Assign yourself as to the PR
- [ ] Assign relevant labels such as `bug`, `enhancement`.
- [ ] Request reviews if the PR is large, complex or you would like an extra pair of eyes to go over it.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] Add new entries to the `Changelog.md`.
- [ ] Update version numbers as needed.

Semver: https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md

https://semver.org/

### CODE REVIEW

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        // An array with the flatened trace table decommitments
+        // For a trace table with n coloums it will be length num_queries*n
+        // and it will be laid out as:
+        // [[query 1 col 1][query 1 col 2]...[query 1 col n]]...[[query q col 1]...[query q col n]]
+        bytes32[] trace_values;
+        // The commitment to those trace values
+        bytes32 trace_root;
+        // The trace table evaluated constraint values at the the query indeicies.
+        bytes32[] constraint_values;
+        // The commitmnet to thos constraint values
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:27]: * those

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        // An array with the public inputs to the STARK
+        bytes32[] public_inputs;
+        // An array with the flatened trace table decommitments
+        // For a trace table with n coloums it will be length num_queries*n
+        // and it will be laid out as:
+        // [[query 1 col 1][query 1 col 2]...[query 1 col n]]...[[query q col 1]...[query q col n]]
+        bytes32[] trace_values;
+        // The commitment to those trace values
+        bytes32 trace_root;
+        // The trace table evaluated constraint values at the the query indeicies.
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:25]: * indices

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    // This function sorts the array
+    // Note - We use insertion sort, the array is expected to be small so this shouldn't
+    // cause problems.
+    function sort(uint64[] memory data) internal pure {
+        for (uint256 i = 0; i < data.length; i++) {
+            uint256 j = i;
+            while (j > 0 && data[j] < data[j - 1]) {
+                uint64 held = data[j];
+                data[j] = data[j - 1];
+                data[j - 1] = held;
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:188]: Solidity has single line swaps I believe. It would look like:

```solidity
(data[j], data[j - 1]) = (data[j], data[j - 1]);
```

Comment (Recmo) [reply crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:188]: And apparently there's a slightly more efficient (not that this currently matters) version:

```
i ← 1
while i < length(A)
    x ← A[i]
    j ← i - 1
    while j >= 0 and A[j] > x
        A[j+1] ← A[j]
        j ← j - 1
    end while
    A[j+1] ← x[3]
    i ← i + 1
end while
```

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            // We don't want a revert if the zero is passed in
+            if (result == 255) {
+                return result;
+            }
+        }
+        return result;
+    }
+
+    // Reads from channel random and returns a list of random queries
+    function get_queries(PublicCoin.Coin memory coin, uint8 max_bit_length, uint8 num_queries)
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:153]: This is part of the FRI algo and should probably get it's own contract

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        // Write the claimed last layer points a set of coeffient for the final layer fri check
+        // NOTE - This is a fri layer so we have to write the whole thing at once
+        coin.write_bytes(abi.encodePacked(proof.last_layer_coeffiencts));
+        return (constraint_coeffiencents, oods_point, oods_coefficients, eval_points);
+    }
+
+    // Given a coin and a nonce hashes the random form the coin and checks that the proof of works passes
+    // NOTE - This function also advances the coin by writing the pow_nonce to it
+    function check_proof_of_work(PublicCoin.Coin memory coin, bytes8 pow_nonce, uint8 pow_bits)
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:123]: Proof of work is a separate sub-protocol. It's cleaner to have it in a separate library so it can be independently tested.

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    struct StarkProof {
+        // An array with the public inputs to the STARK
+        bytes32[] public_inputs;
+        // An array with the flatened trace table decommitments
+        // For a trace table with n coloums it will be length num_queries*n
+        // and it will be laid out as:
+        // [[query 1 col 1][query 1 col 2]...[query 1 col n]]...[[query q col 1]...[query q col n]]
+        bytes32[] trace_values;
+        // The commitment to those trace values
+        bytes32 trace_root;
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:24]: Name is a leaky abstraction: It's a commitment, that it is a merkle *root* is an implementation detail.

Comment (Recmo) [reply crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:24]: Same below for constraint and fri

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        bytes32 constraint_root;
+        // The trace values used for the oods point constraint evaluation
+        bytes32[] trace_oods_values;
+        // The constraint values used for the oods point constraint evaluation
+        bytes32[] constraint_oods_values;
+        // The nonce used for the proof of work
+        bytes8 pow_nonce;
+        // The merkle decomitment for the trace values
+        bytes32[] trace_decommitment;
+        // The merkle decomitment for the constraint evaluted queries
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:37]: * evaluated

File: crypto/stark-verifier-ethereum/test/stark_tests.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    it('It should validate a correct proof', async () => {
+        // NOTE - Typescript has a very very hard time with the ethers js internal array types in struct encoding
+        // in this case it's best for the code to ignore it because this is how ethers js understands these types.
+        // @ts-ignore
+        const return_data = await verifier_contract.verify_proof(testing_proof, constraint_contract.address);
+        expect(return_data);
+    });
+
+    it('It correctly identify the number of leading zeros', async () => {
+        const random_number = utils.bigNumberify('0xfffa751e9886b6eccc725754333f339b7bc9024f38a44e75468fb4b16e1709cc');
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/test/stark_tests.ts:77]: https://www.xkcd.com/221/

File: crypto/stark-verifier-ethereum/contracts/public_coin.sol [hunk 9]
```diff
@@ -1,20 +1,34 @@
...
     // Note that because this is a memory refrence this updates the coin.
-    function write_bytes32(bytes32 to_be_written, Coin memory coin) internal pure {
+    function write_bytes32(Coin memory coin, bytes32 to_be_written) internal pure {
         bytes32 hashed = publicCoinHash(coin.digest, to_be_written);
         coin.counter = 0;
         coin.digest = hashed;
     }
 
+    // Writes a list of bytes32 with each byte written individually
+    function write_many_bytes32(Coin memory coin, bytes32[] memory to_be_written) internal pure {
```
[the remark this answers is not in the document]
Comment (pvienhage) [reply crypto/stark-verifier-ethereum/contracts/public_coin.sol:19]: We use it many times, this is the standard for writing lists of bytes32 to the channel. The lists are only written as groups in the fri layer values.

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        bytes32[] constraint_decommitment;
+        // The extra values needed at each fri layer
```
[the remark this answers is not in the document]
Comment (pvienhage) [reply crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:39]: No they are the extra fri coset values needed to preform fri at each layer.

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol (hunk 3 above)
Comment (pvienhage) [reply crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:188]: I played with implementing the slightly faster version, but it got annoying because of the assumption that j is a signed index and solidity doesn't like that. Decided to just leave the slightly worse one since the difference should very small.

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        bytes32[] fri_commitments;
+        // The merkle proof decommitment at each fri layer
+        bytes32[][] fri_decommitments;
+        // The coeffiencts of the last fri layer
+        bytes32[] last_layer_coeffiencts;
+    }
+
+    // This struct contains the relevent information about the constraint system
+    // It will be returned from a callout to the constraint system contract.
+    struct ConstraintParameters {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:52]: We should rename this struct, it has little to do with the constraints and more with the proof in general. We can call it `ProofParameters`.

File: crypto/stark-verifier-ethereum/test/proof_of_work_test.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    const provider = waffle.provider;
+    const [wallet] = provider.getWallets();
+
+    before(async () => {
+        pow_contract = (await deployContract(wallet, ProofOfWorkTestingArtifact)) as ProofOfWorkTesting;
+    });
+
+    it('Should validate a correct proof of work', async () => {
+        expect(
+            await pow_contract.check_proof_of_work_external(
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/test/proof_of_work_test.ts:24]: Can we also check that it rejects when pow_bits is increased?

File: crypto/stark-verifier-ethereum/test/proof_of_work_test.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+describe('Proof of work testing', () => {
+    let pow_contract: any;
+
+    const provider = waffle.provider;
+    const [wallet] = provider.getWallets();
+
+    before(async () => {
+        pow_contract = (await deployContract(wallet, ProofOfWorkTestingArtifact)) as ProofOfWorkTesting;
+    });
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/test/proof_of_work_test.ts:20]: A test with zero pow_bits would be useful too.

File: crypto/stark-verifier-ethereum/test/testing_proof.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
+import {utils} from 'ethers';
+
+// Note - This proof corresponds to the small fib example in rust.
+// It is not full of accurate data yet.
+
+export const testing_proof: any = {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/test/testing_proof.ts:6]: Can we turn this into a JSON file? You should then be able to import it as:

```
import testing_proof from './testing_proof.json';
```

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol (hunk 1 above)
Comment (Recmo) [reply crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:28]: ^ Still there

File: crypto/stark-verifier-ethereum/contracts/stark_verifier.sol [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        // The roots for each fri decommitment
+        bytes32[] fri_commitments;
+        // The merkle proof decommitment at each fri layer
+        bytes32[][] fri_decommitments;
+        // The coeffiencts of the last fri layer
+        bytes32[] last_layer_coeffiencts;
+    }
+
+    // This struct contains the relevent information about the constraint system
+    // It will be returned from a callout to the constraint system contract.
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/stark_verifier.sol:51]: I wouldn't call this 'relevant information about the constraint system'. Maybe call it configuration for the verifier?

File: crypto/stark-verifier-ethereum/contracts/implementations/trivial_fib.sol [hunk 16]
```diff
@@ -0,0 +1,44 @@
+pragma solidity ^0.6.4;
```
[the remark this answers is not in the document]
Comment (Recmo) [reply crypto/stark-verifier-ethereum/contracts/implementations/trivial_fib.sol:1]: ^ It's not in `testing_contracts`?

File: crypto/stark/src/constraints.rs [hunk 17]
```diff
@@ -82,6 +82,7 @@ impl Constraints {
         if num_reductions % 3 != 0 {
             fri_layout.push(num_reductions % 3);
         }
+        fri_layout = vec![3, 3, 2];
```
Comment (Recmo) [inline crypto/stark/src/constraints.rs]: This would break larger proofs

File: crypto/stark/src/constraints.rs [hunk 18]
```diff
@@ -105,8 +106,8 @@ impl Constraints {
             trace_ncolumns,
             expressions,
             blowup: 16,
-            pow_bits: 0,
-            num_queries: 45,
+            pow_bits: 10,
+            num_queries: 20,
```
Comment (Recmo) [inline crypto/stark/src/constraints.rs]: Why these changes?

File: crypto/stark/src/verifier.rs [hunk 19]
```diff
@@ -210,26 +211,18 @@ pub fn verify(constraints: &Constraints, proof: &Proof) -> Result<()> {
...
         fri_commitments.push(Commitment::from_size_hash(fri_size, &channel.replay())?);
+        eval_points.push(channel.get_random());
     }
-    // Gets the last layer and the polynomial coefficients
-    eval_points.push(channel.get_random());
+    // Gets the last layer coeffiencts
     let last_layer_coefficients = channel.replay_fri_layer(fri_size / constraints.blowup);
 
     // Gets the proof of work from the proof.
+    println!("Iternal coin state {:?}", encode(channel.coin.digest));
```
Comment (Recmo) [inline crypto/stark/src/verifier.rs]: * Internal

File: crypto/stark/src/proof_of_work.rs [hunk 20]
```diff
@@ -8,6 +8,7 @@ use std::sync::atomic::{AtomicU64, Ordering::Relaxed};
 use tiny_keccak::{Hasher, Keccak};
 use zkp_macros_decl::hex;
 use zkp_u256::{Binary, U256};
+use hex::encode;
```
Comment (Recmo) [inline crypto/stark/src/proof_of_work.rs]: Why this change?

File: crypto/stark/src/verifier.rs (hunk 19 above)
Comment (Recmo) [inline crypto/stark/src/verifier.rs]: We shouldn't use `println`, rather `debug!` or `trace!` from the log crate.
